cli/flox-rust-sdk/src/providers/catalog.rs

```diff
@@ -1,12 +1,80 @@
+use std::cell::RefCell;
+use std::collections::VecDeque;
+use std::sync::{Arc, Mutex};
+
 use async_trait::async_trait;
-use catalog_api_v1::types::{self as api_types, error as api_error};
-use catalog_api_v1::{Client as APIClient, Error as APIError};
+use catalog_api_v1::types::{self as api_types, error as api_error, ErrorResponse};
+use catalog_api_v1::{Client as APIClient, Error as APIError, ResponseValue};
 use enum_dispatch::enum_dispatch;
+use reqwest::header::HeaderMap;
+use reqwest::StatusCode;
+use serde::{Deserialize, Serialize};
 use thiserror::Error;
 
 use crate::data::System;
+use crate::models::search::SearchResults;
 
 pub const DEFAULT_CATALOG_URL: &str = "https://flox-catalog.flox.dev";
+pub const FLOX_CATALOG_MOCK_DATA_VAR: &str = "_FLOX_USE_CATALOG_MOCK";
+
+type ResolvedGroups = Vec<ResolvedPackageGroup>;
+
+// Arc allows you to push things into the client from outside the client if necessary
+// Mutex allows you to share across threads (necessary because of tokio)
+// RefCell lets us mutate the field without needing to make the Client trait methods mutable
+type MockField<T> = Arc<Mutex<RefCell<T>>>;
```

Comment:
suggestion: Add comments on how this is used and what "allows you to push things into the client from outside the client" means in this case.
I also have issues figuring out what the scope of a mockfield is supposed to be? 
`read_mock_responses` returns mockfields as a queue/stack of responses, but the type signature represents practically anything?

quesetion/suggestion: Why do we need a `RefCell` here?
My understanding is that `RefCell` allows you get mutable access from a reference _on the same thread_ i.e. a non thread safe `Mutex`. given that we already use a Mutex that gives us metuable references if we need to, adding a refcell to the mix seems a bit redundant 🤔

Reply:
> Add comments on how this is used and what "allows you to push things into the client from outside the client" means in this case.

It allows you to have a handle to the stack of responses without needing to go through the client, which can be useful in testing scenarios.

> I also have issues figuring out what the scope of a mockfield is supposed to be?
`read_mock_responses` returns mockfields as a queue/stack of responses, but the type signature represents practically anything?

This type is meant to be a wrapper around any struct attributes that you'd like to be able to modify on the testing client. Imagine you later add a `next_request_should_fail: bool` field on the client. You'd wrap that in this type as well. Some of the machinery in this wrapper is necessary because we're using the client in an async context.
 
> quesetion/suggestion: Why do we need a `RefCell` here? My understanding is that RefCell allows you get mutable access from a reference on the same thread i.e. a non thread safe Mutex. given that we already use a Mutex that gives us metuable references if we need to, adding a refcell to the mix seems a bit redundant 🤔

That's not the only use case for a `RefCell`. The `*Cell` types provide interior mutability i.e. they allow you to mutate their contents without having a mutable reference or declaring the type as `mut`. One common application is as a wrapper type around a cache. Say you have a client that caches responses. It wouldn't really make sense for the `client.make_a_request` method to always need a `&mut self` parameter just to update the cache, so you make that method take a `&self` and wrap the cache in a `*Cell` type so that you can mutate its contents. I suspect if you look inside the `tracing` crate or any metrics library you'll find a bunch of `*Cell` types, but I haven't looked myself.

`RefCell` in particular let's you get a mutable reference to its contents, as opposed to `Cell` which in most conditions only lets you swap out or set its contents. In this case we want to be able to mutate the contents of the `mock_responses` on the client without needing to change the `Client` trait to always take `&mut self` for the sole purpose of supporting the mock. The `Mutex` is only really necessary because `tokio` requires that types be `Send + Sync`, and a `Mutex<T>` is `Sync` when `T: Send`. In this case `RefCell<T>: Send`.

Reply:
Ah, I see what you mean now. Fixed